server/sonar-server/src/test/java/org/sonar/server/updatecenter/ws/UploadActionTest.java

```diff
@@ -0,0 +1,131 @@
+import org.sonar.server.platform.DefaultServerFileSystem;
+import org.sonar.server.tester.UserSessionRule;
+import org.sonar.server.ws.WsActionTester;
+
+public class UploadActionTest {
+
+  @Rule
+  public TemporaryFolder folder = new TemporaryFolder();
+
+  @Rule
+  public ExpectedException expectedException = ExpectedException.none();
+
+  @Rule
+  public UserSessionRule userSession = UserSessionRule.standalone();
+
+  DefaultServerFileSystem fileSystem = mock(DefaultServerFileSystem.class);
+  File downloadedPluginDir;
+  File plugin = new File(Resources.getResource(getClass(), "UploadActionTest/plugin.jar").getFile());
+
+  @Before
+  public void setUp() throws Exception {
+    downloadedPluginDir = folder.newFolder();
+    when(fileSystem.getDownloadedPluginsDir()).thenReturn(downloadedPluginDir);
+  }
+
+  @Test
+  public void upload_plugin() throws Exception {
+    userSession.login().setGlobalPermissions(SYSTEM_ADMIN);
+
+    newWs().newRequest()
+      .setPart("file", Files.newInputStream(plugin.toPath()), "plugin.jar")
+      .execute();
+
+    Stream<Path> files = Files.list(downloadedPluginDir.toPath());
+    Path result = files.findFirst().get();
+    assertThat(result.toFile().getName()).isEqualTo("plugin.jar");
+  }
+
+  @Test
+  public void erase_existing_plugin_if_already_exists() throws Exception {
+    userSession.login().setGlobalPermissions(SYSTEM_ADMIN);
+
+    newWs().newRequest()
+      .setPart("file", Files.newInputStream(plugin.toPath()), "plugin.jar")
+      .execute();
+
+    // Execute again the WS with the same file
+    newWs().newRequest()
+      .setPart("file", Files.newInputStream(plugin.toPath()), "plugin.jar")
+      .execute();
+
+    Stream<Path> files = Files.list(downloadedPluginDir.toPath());
+    Path result = files.findFirst().get();
+    assertThat(result.toFile().getName()).isEqualTo("plugin.jar");
+  }
+
+  @Test
+  public void fail_when_plugin_extension_is_not_jar() throws Exception {
+    userSession.login().setGlobalPermissions(SYSTEM_ADMIN);
+
+    expectedException.expect(IllegalArgumentException.class);
+    expectedException.expectMessage("Only jar file is allowed");
+    newWs().newRequest()
+      .setPart("file", Files.newInputStream(plugin.toPath()), "plugin.zip")
+      .execute();
+  }
+
+  @Test
+  public void fail_when_no_files_param() throws Exception {
+    userSession.login().setGlobalPermissions(SYSTEM_ADMIN);
+
+    expectedException.expect(IllegalArgumentException.class);
+    expectedException.expectMessage("The 'file' parameter is missing");
+    newWs().newRequest()
+      .execute();
+  }
+
+  @Test
+  public void fail_if_not_system_admin() throws Exception {
+    userSession.login().setGlobalPermissions(PROVISIONING);
+
+    expectedException.expect(ForbiddenException.class);
+    newWs().newRequest()
+      .setPart("file", Files.newInputStream(plugin.toPath()), "plugin.jar")
+      .execute();
+  }
+
+  private WsActionTester newWs(){
```

Comment:
In fact I should have create the WsActionTester in the setup(), just after when(fileSystem.getDownloadedPluginsDir()).thenReturn(downloadedPluginDir);, because we read the downloadedPluginDir in the constructor of UploadAction.